compiler/rustc_parse/src/parser/diagnostics.rs

```diff
@@ -331,7 +331,60 @@ impl<'a> Parser<'a> {
             suggest_raw,
             suggest_remove_comma,
         };
-        err.into_diagnostic(&self.sess.span_diagnostic)
+        let mut err = err.into_diagnostic(&self.sess.span_diagnostic);
+
+        // if the token we have is a `<`
+        // it *might* be a misplaced generic
+        if self.token == token::Lt {
+            // all keywords that could have generic applied
+            let valid_prev_keywords =
+                [kw::Fn, kw::Type, kw::Struct, kw::Enum, kw::Union, kw::Trait];
+
+            // If we've expected an identifier,
+            // and the current token is a '<'
+            // if the previous token is a valid keyword
+            // that might use a generic, then suggest a correct
+            // generic placement (later on)
+            let maybe_keyword = self.prev_token.clone();
+            if valid_prev_keywords.into_iter().any(|x| maybe_keyword.is_keyword(x)) {
+                // if we have a valid keyword, attempt to parse generics
+                // also obtain the keywords symbol
+                match self.parse_generics() {
+                    Ok(generic) => {
+                        if let TokenKind::Ident(symbol, _) = maybe_keyword.kind {
+                            let ident_name = symbol.to_string();
+                            // at this point, we've found something like
+                            // `fn <T>id`
+                            // and current token should be Ident with the item name (i.e. the function name)
+                            // if there is a `<` after the fn name, then don't show a suggestion, show help
+
+                            if !self.look_ahead(1, |t| *t == token::Lt) &&
+                                let Ok(snippet) = self.sess.source_map().span_to_snippet(generic.span) &&
+                                let Ok(ident) = self.sess.source_map().span_to_snippet(self.token.span) {
+                                    err.span_suggestion_verbose(
+                                        generic.span.to(self.token.span),
+                                        format!("place the generic parameter name after the {ident_name} name"),
+                                        format!(" {ident}{snippet}"),
```

Comment:
Could you use a smaller span maybe like this?

```suggestion
                                        self.token.span.shrink_to_hi(),
                                        format!("place the generic parameter name after the {ident_name} name"),
                                        snippet,
```

Example
```
error: expected identifier, found `<`
  --> $DIR/fn-simple.rs:5:3
   |
LL | fn<T> id(x: T) -> T { x }
   |   ^ expected identifier
   |
help: place the generic parameter name after the fn name
   |
LL | fn id<T>(x: T) -> T { x }
   |      ~~~

error: aborting due to previous error
```

Reply:
The suggestion ends up with 

```
error: expected identifier, found `<`
  --> $DIR/fn-simple.rs:5:3
   |
LL | fn<T> id(x: T) -> T { x }
   |   ^ expected identifier
   |
help: place the generic parameter name after the fn name
   |
LL | fn<T> id<T>(x: T) -> T { x }
   |         +++

error: aborting due to previous error
```

notably since we are inserting a suggestion (`<T>`) at the end of `id` - we don't remove the original `<T>` on the right of the `fn`. 

I looked around in the compiler docs, but couldn't quite see a way to do the insertion as before, but able to highlight a portion of the suggestion. The initial PR does the correct suggestion, but I suppose what we want is the original suggestion, but with **only** the insertion of `<T>` after `id` highlighted. 

Thoughts/advice?

Reply:
@TaKO8Ki Just a bump on ^ - In the meantime, I've pushed a revert to the previous span (i.e. the ident and the generic)

Reply:
For this, you'll need a `multipart_suggestion`, which takes a vector of `(Span, String)` pairs. In your case, one of the pair would be `(self.token.span.shrink_to_hi(), snippet)` that @TaKO8Ki suggested, and a second pair `(generics.span, String::new())` to require deletion of the generics.

Reply:
Ooh thanks!! I'll have a look 🎉🎉